src/v/cluster/tm_tx_hash_ranges.h

```diff
@@ -0,0 +1,301 @@
+        if (!hash_ranges.contains(hash)) {
+            return tm_tx_hash_ranges_errc::not_hosted;
+        }
+        excluded_transactions.insert(tx_id);
+        return tm_tx_hash_ranges_errc::success;
+    }
+
+    tm_tx_hash_ranges_errc
+    include_transaction(const kafka::transactional_id& tx_id) {
+        if (included_transactions.contains(tx_id)) {
+            return tm_tx_hash_ranges_errc::success;
+        }
+        if (excluded_transactions.contains(tx_id)) {
+            excluded_transactions.erase(tx_id);
+            vassert(
+              hash_ranges.contains(get_tx_id_hash(tx_id)),
+              "hash_ranges must not contain excluded txes");
+            return tm_tx_hash_ranges_errc::success;
+        }
+        uint32_t hash = get_tx_id_hash(tx_id);
+        if (hash_ranges.contains(hash)) {
+            return tm_tx_hash_ranges_errc::intersection;
+        }
+        included_transactions.insert(tx_id);
+        return tm_tx_hash_ranges_errc::success;
+    }
+
+    tm_tx_hash_ranges_errc add_range(tm_hash_range range) {
+        if (hash_ranges.intersects(range)) {
+            return tm_tx_hash_ranges_errc::intersection;
+        }
+        for (const auto& tx_id : excluded_transactions) {
+            uint32_t hash = get_tx_id_hash(tx_id);
+            if (range.contains(hash)) {
+                excluded_transactions.erase(tx_id);
+            }
+        }
+        for (const auto& tx_id : included_transactions) {
+            uint32_t hash = get_tx_id_hash(tx_id);
+            if (range.contains(hash)) {
+                included_transactions.erase(tx_id);
+            }
+        }
+        hash_ranges.add_range(range);
+        return tm_tx_hash_ranges_errc::success;
+    }
+
+    bool contains(const kafka::transactional_id& tx_id) {
+        if (excluded_transactions.contains(tx_id)) {
+            return false;
+        }
+        if (included_transactions.contains(tx_id)) {
+            return true;
+        }
+        auto tx_id_hash = get_tx_id_hash(tx_id);
+        return hash_ranges.contains(tx_id_hash);
+    }
+};
+
+} // namespace cluster
+
+namespace reflection {
+
+template<>
+struct adl<cluster::tm_hash_range> {
+    void to(iobuf& out, cluster::tm_hash_range&& hr) {
+        reflection::serialize(out, hr.first, hr.last);
+    }
+    cluster::tm_hash_range from(iobuf_parser& in) {
+        auto first = reflection::adl<uint32_t>{}.from(in);
+        auto last = reflection::adl<uint32_t>{}.from(in);
+        return {first, last};
+    }
+};
+
+template<>
+struct adl<cluster::tm_hash_ranges_set> {
+    void to(iobuf& out, cluster::tm_hash_ranges_set&& hr) {
+        reflection::serialize(out, hr.ranges);
+    }
+    cluster::tm_hash_ranges_set from(iobuf_parser& in) {
+        auto ranges
+          = reflection::adl<std::vector<cluster::tm_hash_range>>{}.from(in);
+        return {std::move(ranges)};
+    }
+};
+
+template<>
+struct adl<cluster::tm_tx_hosted_transactions> {
```

Comment:
is this new usage of adl (rather than serde?)

Reply:
Roman switched to serde to serialized a new batch. This adl is used to update a snapshot which is adl based. We need to switch to the snapshot serialization to serde at some point but it requires a thought and isn't directly relevant to this PR.

Reply:
thanks, makes sense. @ZeDRoman please make sure that context like this is discoverable by future developers without having to re-reading the PR discussion. you can literally write as much prose as you want in the commit message.